lahja/endpoint.py

```diff
@@ -508,30 +712,30 @@ def subscribe(self,
         of events was received.
         """
         queue: 'asyncio.Queue[TStreamEvent]' = asyncio.Queue()
-
+        casted_queue = cast('asyncio.Queue[BaseEvent]', queue)
         if event_type not in self._queues:
             self._queues[event_type] = []
 
-        # mypy cannot reconcile BaseEvent with TStreamEvent
-        self._queues[event_type].append(queue)  # type: ignore
-        i = None if num_events is None else 0
-        while True:
-            try:
-                yield await queue.get()
-            except (GeneratorExit, asyncio.CancelledError):
-                # mypy cannot reconcile BaseEvent with TStreamEvent
-                self._queues[event_type].remove(queue)  # type: ignore
-                break
-            else:
-                if i is None:
-                    continue
+        self._queues[event_type].append(casted_queue)
+        await self._notify_subscriptions_changed()
 
-                i += 1
+        if num_events is None:
+            # loop forever
+            iterations = itertools.repeat(True)
+        else:
+            iterations = itertools.repeat(True, num_events)
 
-                if i >= cast(int, num_events):
-                    # mypy cannot reconcile BaseEvent with TStreamEvent
-                    self._queues[event_type].remove(queue)  # type: ignore
+        try:
+            for _ in iterations:
+                try:
+                    yield await queue.get()
+                except GeneratorExit:
+                    break
+                except asyncio.CancelledError:
                     break
+        finally:
+            self._queues[event_type].remove(casted_queue)
```

Comment:
You might take a look at my `trio` implementation which uses a `weakref.WeakSet` to store the queues so that this cleanup step can be omitted.